cosmoz-image-viewer.js

```diff
@@ -344,52 +352,123 @@
+				return this._detachToExistingWindow();
 			}
 
+			globals.windowBeforeUnloadHandler = () => {
+				if (globals.windowOpener !== null) {
+					globals.windowOpener._setIsDetached(false);
+				}
+				globals.windowOpener = null;
+				globals.window = null;
+			};
+
+			return this._detachToNewWindow();
+		},
+
+		_detachToNewWindow() {
 			const
 				w = globals.window = window.open(undefined, this.detachedWindowName, this._detachedWindowFeaturesString),
 				windowTemplate = this.$$('#externalWindow'),
-				windowTemplateClone = windowTemplate.content.cloneNode(true);
+				windowTemplateClone = windowTemplate.content.cloneNode(true),
+				setImages = () => w.ciw.setImages(this._resolvedImages, this.currentImageIndex);
 
 			if (w == null) {
 				// if window.open() is blocked (popup blocked, not emited by native user triggered event)
-				return;
+				return w;
 			}
 
 			if (w.ciw != null) {
 				// browser has been refreshed without closing external window, we will reuse an old window
 				w.removeEventListener('beforeunload', globals.windowBeforeUnloadHandler);
 			}
 
-			globals.windowBeforeUnloadHandler = () => {
-				globals.windowOpener._setIsDetached(false);
-				globals.windowOpener = null;
-				globals.window = null;
-			};
-
 			w.document.title = this._detachedWindowTitle;
 
+			w.addEventListener('download', ({detail}) => this.createZipFromUrls(detail).then(zip => this.downloadZip(zip)));
 			w.addEventListener('beforeunload', globals.windowBeforeUnloadHandler);
 
 			if (w.ciw == null) {
-				w.addEventListener('ready', () => w.ciw.setImages(this._resolvedImages, this.currentImageIndex));
-				w.document.body.appendChild(windowTemplateClone);
+				w.addEventListener('ready', () => setImages());
+
+				Array.from(windowTemplateClone.childNodes)
+					.forEach(node => {
+						if (node.tagName === 'SCRIPT') {
+							// Needed for Firefox
+							// otherwise the script would not be evaluated
+							const sc = document.createElement('script');
+							sc.innerHTML = node.innerHTML;
+							w.document.body.appendChild(sc);
+							return;
+						}
+						w.document.body.appendChild(node);
+					});
 			} else {
-				w.ciw.setImages(this._resolvedImages, this.currentImageIndex);
+				setImages();
 			}
+
+			return w;
 		},
+
+		_detachToExistingWindow() {
+			globals.window.document.title = this._detachedWindowTitle;
+			globals.window.ciw.setImages(this._resolvedImages, this.currentImageIndex);
+			globals.window.focus();
+			return globals.window;
+		},
+
 		get hasWindow() {
-			return globals.window != null && !globals.window.closed;
+			return globals.window != null && !globals.window.closed && globals.window.ciw;
 		},
+
 		syncState() {
 			if (!this.isDetached && this.hasWindow) {
 				this.detach();
 			}
 		},
+
+		downloadZip(zip) {
+			const a = document.body.appendChild(zip.createDownloadLink());
+			a.click();
+			document.body.removeChild(a);
+		},
+
+		createZipFromUrls(fileUrls) {
+			const fetches = fileUrls
```

Comment:
I feel the `fetch` call is somewhat obfuscated with this indentation.

How about

```js
const fetches = fileUrls.map(url =>
  fetch(url)
    .then(response => response.arrayBuffer())
    .then(data => ({data, url}))
);
```

Reply:
sweet!